### DISCUSSION TITLE
Translating method shorhand syntax

### DISCUSSION THREAD

**Discussion by char0n** [babel/babel#12874] [Q&A]
I have a special corner case during method shorthand transpilation. I have a visitor mechanism that consumes visitors as defined below.

```js
const visitor = {
  Object() {
  },
  Array() {
  },
}
```

Following visitor is transpiled into:

```js
"use strict";

var visitor = {
  Object: function Object() {},
  Array: function Array() {}
};
```
As function declarations are hoisted, functions `Object` and `Array` overrides global `Array` and `Object` objects. This causes a lot of problems as you can imagine. 

Is there a sensible workaround/solution for this?

I currently see only following options: 
 - use different mappings for my visitors
 - create a transformation plugin for babel to rename the function names to `_Object` and `_Array`.
 
 Thank you for any help!

**Comment by char0n** [comment]
Hm when assigned to an object property it seems that it doesn't override global objects. My problem is somewhere else it seems.

**Reply by char0n** [reply]
It seem to be preblamatic when combining with webpack:

```js
const visitor = { 
  Object(element) {
    isObjectElement(element);
    return BREAK;
  } 
}
```

gets translated with babel into:

```js
"use strict";

var visitor = {
  Object: function Object(element) {
    isObjectElement(element);
    return BREAK;
  }
};
```
which gets translated by webpack to

```js
{
    Object: function Object(element) {
      Object(moduleX__WEBPACK_IMPORTED_MODULE_2__["isObjectElement"])(element);
      return moduleX__WEBPACK_IMPORTED_MODULE_2__["BREAK"];
    }
}
```

As we can see here, with our shorthand translation to `function Object` we override the inner use of buildin `Object` function.

Babel has a prevention for this:
```js
const a = {
  Object(element) {
    console.dir(Object(element));
  },
}
```
translates to

```js
"use strict";

var a = {
  Object: function (_Object) {
    function Object(_x) {
      return _Object.apply(this, arguments);
    }

    Object.toString = function () {
      return _Object.toString();
    };

    return Object;
  }(function (element) {
    console.dir(Object(element));
  })
};
```
I  assume it detects that `Object` has been used inside the `Object` method.

**Reply by nicolo-ribaudo** [reply]
Good investigation! Yes, Babel makes sure that it doesn't introduce conflicts. As you pointed out, this is a webpack bug: it should check that `Object` isn't shadowed.

However, there are two workarounds you can use to prevent Babel from generating that code that then breaks webpack:

1. Pass `exclude: ["transform-function-name"]` to your `@babel/preset-env`'s options. By doing so it shouldn't add the `Object` name to the function.
2. Don't use object methods for `Object`. You cold directly write
    ```js
    const visitor = {
      Object: function _Object() {
      },
      Array() {
      },
    }
    ```

**Comment by char0n** [comment]
@nicolo-ribaudo,

Thanks for suggestions.  2.) is a **nogo** as there are multiple ways how to define visitor methods and I just don't like to enforce it.

For now I'll try option 1.). I'm currently using CRA which uses webpack@4. I'll have to check if this still manifests in webpack@5 and I'll report the bug there (if it still exists).

**Reply by char0n** [reply]
Only webpack@4 is affected. webpack@5 works correctly.

**Reply by char0n** [reply]
Created bug issue for webpack@4: https://github.com/webpack/webpack/issues/12767
